mdp/signal_node_online.py

```diff
@@ -0,0 +1,376 @@
+        data can be processed forwards (or backwards by applying the inverse
+        of the transformation computed by the node if defined).
+
+        Unlike a Node, an OnlineNode's execute (or inverse) call __does not__ end the
+        training phase of the node. The training phase can only be stopped through
+        an explicit 'stop_training' call. Once stopped, an OnlineNode becomes
+        functionally equivalent to a trained Node.
+
+        OnlineNode also supports multiple training phases. However, all
+        the training_phases are active for each data point. See _train_seq
+        documentation for more information.
+
+        The training type of an OnlineNode can be set either to 'incremental'
+        or 'batch'. When set to 'incremental', the input data array is passed for
+        training sample by sample, while for 'batch' the entire data array is passed.
+
+        An `OnlineNode` inherits all the Node's utility methods, for example
+        `copy` and `save`, that returns an exact copy of a node and saves it
+        in a file, respectively. Additional methods may be present, depending
+        on the algorithm.
+
+        OnlineNodes also support using a pre-seeded random number generator
+        through a 'numx_rng' argument. This can be useful to replicate
+        results.
+
+        `OnlineNode` subclasses should take care of overwriting (if necessary)
+        the functions `_train`, `_stop_training`, `_execute`, 'is_trainable',
+        `is_invertible`, `_inverse`, `_get_supported_dtypes` and '_get_supported_training_types'.
+        If you need to overwrite the getters and setters of the
+        node's properties refer to the docstring of `get_input_dim`/`set_input_dim`,
+        `get_output_dim`/`set_output_dim`, `get_dtype`/`set_dtype`, 'get_numx_rng'/'set_numx_rng'.
+    """
+
+    def __init__(self, input_dim=None, output_dim=None, dtype=None, numx_rng=None):
+        """If the input dimension and the output dimension are
+        unspecified, they will be set when the `train` or `execute`
+        method is called for the first time.
+        If dtype is unspecified, it will be inherited from the data
+        it receives at the first call of `train` or `execute`.
+        Every subclass must take care of up- or down-casting the internal
+        structures to match this argument (use `_refcast` private
+        method when possible).
+        If numx_rng is unspecified, it will be set to a random number generator
+        with a random seed.
+        """
+        super(OnlineNode, self).__init__(input_dim, output_dim, dtype)
+        # this var stores the index of the current training iteration
+        self._train_iteration = 0
+        # this cache var dict stores an interemediate result or paramenter values
+        # at the end of each train iteration. OnlineNode subclasses should
+        # initialize the required keys
+        self._cache = dict()
```

Comment:
What is the advantage of having the cache dict, instead of dedicated variables in subclasses?
Maybe I missed it, but I don't see any general functionality that is enabled by this?